Dm events

agrover opened the pull request:
pending dm-rs changes. WIP.

agrover commented:
`event_on()` only placeholder code.

agrover commented:
Rebased.
* Added some missing trait method docstrings

mulkieran commented on src/bin/stratisd.rs:
```diff
@@ -109,8 +133,13 @@ fn run() -> StratisResult<()> {
+        // Refresh list of dbus fds to poll for every time. This can change as
+        // D-Bus clients come and go.
+        fds.truncate(dbus_fd_start_idx);
+        fds.extend(dbus_conn.watch_fds().iter().map(|w| w.to_pollfd()));
```
In this new arrangement, the first time through the loop there are no d-bus events to be interested in, but there were before. Is that intentional?

agrover replied:
The loop is structured to avoid duplicating code. There are things that need to happen before `poll()` is called the first time (setup), and then there are things (checking, setup) every time `poll()` wakes up. That ends up putting the call to `poll()` at the bottom of the loop, so the first time through the "checking" part is moot, we're just doing the "setup" part... and then in subsequent loops both checking and setup are done.

mulkieran commented on src/bin/stratisd.rs:
```diff
@@ -109,8 +133,13 @@ fn run() -> StratisResult<()> {
+        fds.truncate(dbus_fd_start_idx);
+        fds.extend(dbus_conn.watch_fds().iter().map(|w| w.to_pollfd()));
+
+        let r = unsafe { libc::poll(fds.as_mut_ptr(), fds.len() as libc::c_ulong, poll_timeout) };
```
Why does whether or not engine has an eventable determine timeout for all interesting file descriptors?

agrover replied:
Because if the engine doesn't support events then it needs to poll -- periodically have a chance to check itself to see if it needs to do something. (We were doing this before for StratEngine to call `check()`, which would e.g. extend the flex data device if needed, etc..)

mulkieran replied:
I should be more specific...most of those fd's have nothing to do with the engine's eventables. Why is their poll-timeout dependent on the existance of the eventable when it wasn't before?

agrover replied:
I don't follow -- there's just one timeout for all the fds we're polling. The other fds, the D-Bus fds, are already event-driven so if any of them have activity, we return from poll() immediately and handle them. Once we can avoid polling by making the engine check() calls event-driven, then there's no need to "fake" a DM event every 10 seconds, so the timeout serves no purpose any more.

mulkieran commented on src/bin/stratisd.rs:
```diff
@@ -109,8 +133,13 @@ fn run() -> StratisResult<()> {
-        engine.borrow_mut().check()
+        // Refresh list of dbus fds to poll for every time. This can change as
+        // D-Bus clients come and go.
+        fds.truncate(dbus_fd_start_idx);
```
Just set up dbus fds here (rather than adding to uber-list of fds.

agrover replied:
We can't do this because it all needs to be in the uber-list.

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -122,6 +141,42 @@ impl Engine for StratEngine {
         self.pools.into_iter().map(|x| x as &Pool).collect()
     }
+
+    fn get_eventable(&mut self) -> EngineResult<Option<Box<Eventable>>> {
```
I'm not sure that Box in the actual return type is a good idea.

agrover replied:
We need to return a trait object here, so I think this is needed.

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -25,6 +27,20 @@ pub enum DevOwnership {
 #[derive(Debug)]
 pub struct StratEngine {
     pools: Table<StratPool>,
+    watched_devs: HashMap<DmNameBuf, u32>, // name, event_nr
```
Needs a comment on new field, something like:

```// Maps name of DM device to most recent event number associated w/ the device```

agrover replied:
ok

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -122,6 +141,42 @@ impl Engine for StratEngine {
+                                break;
+                            }
+                        }
+                        *prev_event_nr = new_event_nr;
```
Given that ```watched_devs```is reset just below, I'm not sure that this assignment achieves anything.

agrover replied:
oh yeah! ok.

mulkieran commented on src/engine/strat_engine/pool.rs:
```diff
@@ -117,6 +117,19 @@ impl StratPool {
+    }
+
+    pub fn event_on(&mut self, dm_name: &DmName) -> EngineResult<()> {
+        println!("event on {}", dm_name);
```
Expecting this to go away, of course.

agrover replied:
yup.

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -122,6 +141,42 @@ impl Engine for StratEngine {
+            }
+        }
+
+        self.watched_devs = device_list
```
I don't thing these devs are really watched. Maybe ```last_observed_event_nrs``` would be a better name.

agrover replied:
yes, ok

agrover replied:
On second thought, they are. This is a list of only the dm names we care about, and the last seen event numbers for each. I'll try to name it something better, and we can see if it's better.

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -122,6 +141,42 @@ impl Engine for StratEngine {
+        let device_list = DM::new()?.list_devices()?;
+
+        for &(ref dm_name, _, new_event_nr) in &device_list {
+            if let Some(new_event_nr) = new_event_nr {
```
I think you can drop both these if's. Instead,

```if new_event_nr > self.watched_devs.get_mut(dm_name) {```

should work out correctly.

agrover replied:
I think there was a theoretical concern about event_nr wrapping?

mulkieran replied:
Actually, I think there may be a bug in your current formulation. I think that if ```watched_devs``` lacks an entry for ```dm_name``` but there was a corresponding ```event_nr``` for the name in the device list then you had better take the action. But in your current code, you don't.

agrover replied:
If watched_devs lacks an entry, then an event happened on a DM device we don't care about. DM::list_devices() returns all DM devices on the system. If e.g. an LVM configured DM device has an event we will still end up in this function and should take no action.

mulkieran replied:
I don't think that's quite right.

Last time we called ```evented()```, a particular device might not have existed and ```watched_dev_last_event_nrs``` would not contain a key for that device. But this time we call ```evented()``` the device does exist, so it is contained in ```list_devices()``` result but not in ```watched_dev_last_event_nrs```. I think we need to do something in that case.

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -122,6 +141,42 @@ impl Engine for StratEngine {
+                    if new_event_nr > *prev_event_nr {
+                        // Event! We know the dm name, but don't know which
+                        // pool
+                        for pool in &mut self.pools {
```
Might as well use iterator ```find()``` method here.Or possibly ```position``` and then ```get_mut()``` if found.

agrover replied:
ok will try.

agrover replied:
got this working with find(), but it's ugly:
```rust
if let Some(pool) = (&mut self.pools)
                               .into_iter()
                               .find(|ref mut pool| {
                                         pool.get_eventing_dev_names().contains(dm_name)
                                     }) {
                            pool.event_on(dm_name)?;
                        }
```

I think working on #492 might help, not sure. Maybe we can go with what's there for now and then circle back.

agrover commented on src/engine/engine.rs:
```diff
@@ -172,4 +173,20 @@ pub trait Engine: Debug {
+
+    /// If the engine would like to include an event in the message loop, it
+    /// may return an Eventable from this method.
+    fn get_eventable(&mut self) -> EngineResult<Option<Box<Eventable>>>;
```
We can do this when such a thing becomes necessary.

mulkieran replied:
I honestly think it might make the funky code up above easier. But, I'm happy to leave it for now.

agrover commented on src/engine/engine.rs:
```diff
@@ -172,4 +173,20 @@ pub trait Engine: Debug {
+    /// may return an Eventable from this method.
+    fn get_eventable(&mut self) -> EngineResult<Option<Box<Eventable>>>;
+
+    /// Notify the engine that an event has occurred on the Eventable.
```
That's all we need for StratEngine right now. It's really hard to know what we'll need in the future, and easy enough to add in the future as well, so this is just doing what's required for today's situation where we only have Sim and StratEngine.

agrover commented on src/engine/strat_engine/thinpool.rs:
```diff
@@ -543,6 +542,14 @@ impl ThinPool {
         }
     }
+
+    pub fn get_eventables(&self) -> Vec<DmNameBuf> {
```
I don't see how to avoid making this Vec without making things more complex.

agrover commented on src/bin/stratisd.rs:
```diff
@@ -85,20 +85,44 @@ fn run() -> StratisResult<()> {
+    let mut fds = Vec::new();
+
+    let mut eventable = engine.borrow_mut().get_eventable()?;
+    let poll_timeout = match eventable {
```
You have to eventually make one list of `struct pollfd` because that's what `libc::poll` takes, see `man 2 poll`. And then that's also where you look after it returns to tell which (if any) of the fds have had an event. I'm not sure doing as you describe simplifies things -- so we now have an `Option<libc::pollfd>` but you still have to copy it into a Vec with the dbus fds, one after the other, and then check the array to dispatch events. How can this be made simpler? There seems to be a lot of irreducible ugliness here (which is what things like tokio is supposed to improve, and maybe we should use that in the future, but it has its own set of complexities that IMHO make this simpler right now.)

agrover commented on src/bin/stratisd.rs:
```diff
@@ -85,20 +85,44 @@ fn run() -> StratisResult<()> {
+        // missed but may result in an unneeded call to evented().
+        match eventable {
+            Some(ref mut evt) => {
+                if fds[0].revents != 0 {
```
We can't do this because it all needs to be in the uber-list.

agrover commented:
rebased.
* Added 2 squashable commits that rename & add comment for clarity

mulkieran commented on src/bin/stratisd.rs:
```diff
@@ -85,20 +85,44 @@ fn run() -> StratisResult<()> {
+    let mut fds = Vec::new();
+
+    let mut eventable = engine.borrow_mut().get_eventable()?;
+    let poll_timeout = match eventable {
```
```
if let Some(ref mut evt) = eventable {
  fds.push(...)
}
let dbus_fd_start_idx = fds.len();
```

mulkieran commented on src/bin/stratisd.rs:
```diff
@@ -109,8 +133,13 @@ fn run() -> StratisResult<()> {
+        fds.truncate(dbus_fd_start_idx);
+        fds.extend(dbus_conn.watch_fds().iter().map(|w| w.to_pollfd()));
+
+        let r = unsafe { libc::poll(fds.as_mut_ptr(), fds.len() as libc::c_ulong, poll_timeout) };
```
instead of ```poll_timeout```,

```if dbus_fd_start_idx == 0 { 10000 } else { -1 }```

mulkieran commented on src/engine/strat_engine/thinpool.rs:
```diff
@@ -543,6 +542,14 @@ impl ThinPool {
         }
     }
+
+    pub fn get_eventing_dev_names(&self) -> Vec<DmNameBuf> {
```
I guess method is probably fine. Suggested header comment:

```
/// The names of DM devices belonging to this pool that may generate events
```

?

mulkieran commented on src/engine/strat_engine/pool.rs:
```diff
@@ -117,6 +117,19 @@ impl StratPool {
         self.thin_pool.has_filesystems()
     }
+
+    pub fn get_eventing_dev_names(&self) -> Vec<DmNameBuf> {
```
This method probably also needs a header comment like the one below.

agrover commented:
rebased.
* Add comments to get_eventing_dev_names functions
* Redo code around poll_timeout and dbus_fd_start_idx to make it clearer
      their relative importance

mulkieran commented on src/bin/stratisd.rs:
```diff
@@ -85,20 +85,45 @@ fn run() -> StratisResult<()> {
-        // Poll them with a 10 s timeout
-        let r = unsafe { libc::poll(fds.as_mut_ptr(), fds.len() as libc::c_ulong, 10000) };
-        assert!(r >= 0);
+        // Handle all Eventable events before polling. If an event arrives
```
You're only handling one or zero Eventable events in this code. It doesn't generalize.

agrover replied:
Ok I'll clarify. We can have multiple DM events that result in a single Eventable, but at this level that is not relevant.

agrover commented:
> Why do we need to return a trait object? Is it due to unknown size?

Yes. We know we have (a reference to) a struct that implements Eventable, but we don't know anything else about the struct, including its size.

agrover commented:
* fix comment about Eventable

mulkieran commented on src/bin/stratisd.rs:
```diff
@@ -109,8 +134,13 @@ fn run() -> StratisResult<()> {
+        fds.extend(dbus_conn.watch_fds().iter().map(|w| w.to_pollfd()));
+
+        let r = unsafe { libc::poll(fds.as_mut_ptr(), fds.len() as libc::c_ulong, poll_timeout) };
+        assert!(r >= 0);
```
I think ```assert!``` is probably not the write way to check ```poll()``` return code.

agrover replied:
We really shouldn't be using `libc::poll` at all. I've opened #655, which will include removing that assert.

mulkieran commented on src/bin/stratisd.rs:
```diff
@@ -85,20 +85,45 @@ fn run() -> StratisResult<()> {
-        // Poll them with a 10 s timeout
-        let r = unsafe { libc::poll(fds.as_mut_ptr(), fds.len() as libc::c_ulong, 10000) };
-        assert!(r >= 0);
+        // Handle the Eventable (if there is one) before polling. If an event
```
This code still seems very funky, but I feel as if I understand it.

agrover replied:
it is funky. At some point we may want to look at mio or tokio crates to abstract the messiness of the event loop. I've opened #656 to track.

mulkieran commented on src/engine/engine.rs:
```diff
@@ -172,4 +173,20 @@ pub trait Engine: Debug {
+    fn evented(&mut self) -> EngineResult<()>;
+}
+
+pub trait Eventable {
```
Please document the purpose of the trait.

agrover commented:
squashed and rebased.

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -122,6 +143,41 @@ impl Engine for StratEngine {
+    fn evented(&mut self) -> EngineResult<()> {
+        let device_list = DM::new()?.list_devices()?;
+
+        for &(ref dm_name, _, new_event_nr) in &device_list {
```
Either we'll be using a newer version of devicemapper, i which case every single one of these ```new_event_nr``` values will be ```Some``` or we'll not, in which case they will all be ```None```. In the second case, no pool will ever be "evented on".

We should definitely document what our stance is on that situation.

agrover replied:
Yeah this is a problem. Or at least a weirdness. Maybe if `None` we should instead *always* call event_on() for all watched devs?

mulkieran replied:
I think that if devicemapper version is early so that no event nrs available, then the only meaningful thing is to call ```event_on()``` for all pools.

I think that in that same case, there will be no watched devs.

agrover commented:
> I thin nix::poll will have the same meaning in its return value, though, so it might still be worth handling here.

libc::poll returns a c_int, but nix::poll returns a `Result<c_int>`, so it will be different in that the -1 that libc::poll returns to indicate error an error is checked-for by Nix, and results in an `Err` value. I think with nix::poll we can just `?` it, since we don't happen to care about the value (the number of fds with events) as long as there was no error.

mulkieran commented:
> > I thin nix::poll will have the same meaning in its return value, though, so it might still be worth handling here.

> libc::poll returns a c_int, but nix::poll returns a `Result<c_int>`, so it will be different in that the -1 that libc::poll returns to indicate error an error is checked-for by Nix, and results in an `Err` value. I think with nix::poll we can just `?` it, since we don't happen to care about the value (the number of fds with events) as long as there was no error.

Ok!

agrover commented:
* Added minimum DM version check to StratEngine and simplified `evented()` implementation now that we can assume list_devices() always returns an event_nr.

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -35,6 +55,14 @@ impl StratEngine {
     /// Returns an error if there was an error reading device nodes.
     /// Returns an error if there was an error setting up any of the pools.
     pub fn initialize() -> EngineResult<StratEngine> {
+        let minor_dm_version = DM::new()?.version()?.1;
```
We might want to document the reason for this requirement in some docs.

agrover replied:
opened https://github.com/stratis-storage/stratis-docs/issues/72

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -122,6 +153,40 @@ impl Engine for StratEngine {
+
+        self.watched_dev_last_event_nrs = device_list
+            .into_iter()
+            .filter_map(|(dm_name, _, event_nr)| match event_nr {
```
There's no need to filter here anymore, since ```even_nr``` can not be ```None```.

agrover replied:
ok

agrover commented:
rebased.
* Don't need `filter_map`, just use `map`
* Add a comment line to StratEngine::initialize()

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -122,6 +154,39 @@ impl Engine for StratEngine {
+        let device_list = DM::new()?.list_devices()?;
+
+        for &(ref dm_name, _, new_event_nr) in &device_list {
+            let new_event_nr = new_event_nr.expect("Supported DM versions always provide a value");
```
Now that we know that ```new_event_nr``` is always ```Some``` this is simpler to think about.

I still think that

```
if new_event_nr > self.watched_dev_last_event_nrs.get(dm_name) {
  ...
}
```
is the correct thing to do.

There are only two possiblites.

1. ```prev_event_nr``` is ```None```. In that case, the device has shown up since the last time ```evented()``` was called. So there may be something to do about it, and we should handle it.
2. ```prev_event_nr``` matches ```Some```. In that case, there will be a comparsion between the values of the two numbers and if ```new_event_nr``` is greater than ```prev_event_nr``` something will be done.

agrover replied:
Doesn't compile.
```
error[E0308]: mismatched types
   --> src/engine/strat_engine/engine.rs:167:31
    |
167 |             if new_event_nr > self.watched_dev_last_event_nrs.get(dm_name) {
    |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected u32, found enum `std::option::Option`
    |
    = note: expected type `u32`
               found type `std::option::Option<&u32>`
```

mulkieran replied:
Yeah, you need to lose the ```expect()``` call as well. Then they are both ```Option``` type and hopefully ```Option``` has enough cleverness that ```>``` will still work.

mulkieran replied:
But in any case, if you do nothing, I think you have  a bug.

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -122,6 +154,39 @@ impl Engine for StratEngine {
+                if new_event_nr > *prev_event_nr {
+                    // Event! We know the dm name, but don't know which
+                    // pool
+                    for pool in &mut self.pools {
```
Was it impossible to use ```find``` here?

agrover replied:
no, but it looks like this:
```
if let Some(pool) =
    (&mut self.pools)
        .into_iter()
        .find(|pool| pool.get_eventing_dev_names().contains(dm_name)) {
    pool.event_on(dm_name)?;
}
```

which doesn't seem much better than what's there now.

mulkieran commented on src/engine/strat_engine/pool.rs:
```diff
@@ -117,6 +117,19 @@ impl StratPool {
+    }
+
+    pub fn event_on(&mut self, _dm_name: &DmName) -> EngineResult<()> {
+        // We could be smart and just check the thing that evented, but for
```
Could you turn this into a ```TODO``` or a ```FIXME```, please?

agrover replied:
ok

mulkieran commented on src/engine/strat_engine/pool.rs:
```diff
@@ -117,6 +117,19 @@ impl StratPool {
+    pub fn event_on(&mut self, _dm_name: &DmName) -> EngineResult<()> {
+        // We could be smart and just check the thing that evented, but for
+        // now check everything
+        self.check()?;
```
Single line, ```self.check()``` would be more aesthetic, but this is not important.

agrover replied:
sure, why not

agrover commented:
* made style changes to event_on()

mulkieran commented on src/bin/stratisd.rs:
```diff
@@ -85,20 +85,45 @@ fn run() -> StratisResult<()> {
-        let r = unsafe { libc::poll(fds.as_mut_ptr(), fds.len() as libc::c_ulong, 10000) };
-        assert!(r >= 0);
+        // Handle the Eventable (if there is one) before polling. If an event
+        // arrives after calling clear_event(), we will pop out of the poll
```
most of this comment is obscure to me. "If an event arrives after calling clear_event()``` is the unclearest to me. I'm not sure we do guarantee that no events are missed. We could miss 100 and then detect the 101st, AFAICT.

agrover replied:
Can you give an example of scenario where you think we would miss an event? Then I can walk through it.

mulkieran replied:
Well I'm assuming a one-to-one and onto relationship between events and event nrs. Suppose you call list devices. You look at all devices and the highest event_nr is n. Later you call list_devices again. The device which previously had event_nr n now has event_nr n + 2. All other devices had the same event nr they had previously. I don't think that is impossible, but that means that you missed event corresponding to event nr n + 1.

agrover replied:
Yes, that is certainly possible. That's working as intended. The goal is to get woken up when an event happens, and then know which DM device(s) evented (once or more) times since the last time we handled events. The assumption is we'd get the current status of the device and take action based on the status. If the status shows multiple things needing attention with the device then we should resolve all of them, rather than getting a separate event for each issue.

agrover commented:
rebased.
* Rewrite evented() to handle if we don't know the last seen event for a watched dev

mulkieran commented on src/engine/strat_engine/engine.rs:
```diff
@@ -122,6 +154,41 @@ impl Engine for StratEngine {
+
+        for pool in &mut self.pools {
+            for dm_name in pool.get_eventing_dev_names() {
+                if let Some(new_event_nr) = device_list.get(&dm_name) {
```
```
if device_list.get(&dm_name) > self.watched_dev_last_event_nrs.get(&dm_name) {
  pool.even_on(&dm_name)?;
}
```

should still take careof lines 173 - 185, AFAICT.

agrover commented:
ooh that's slick!

Made that change, rebased, and squashed.

mulkieran commented:
I prefer the arrangement shown here: https://github.com/mulkieran/stratisd/commit/9cefb5f9efe4b9bae26ad9198755ac1c1646ea08. It gives a uniformity of structure as now the code works over the list. Also, it makes the relation between the idx and existance of eventable more explicit.

mulkieran commented:
Actually, I prefer the whole approach in: https://github.com/mulkieran/stratisd/commits/agrover-dm-events.

mulkieran commented:
Also, I'm not convinced that ```poll_timeout``` conditional is not reversed. Please verify that its correct and please add an explanatory comment.

mulkieran commented:
I'm gonna try one more thing...

mulkieran commented:
Probably the best available: https://github.com/mulkieran/stratisd/tree/agrover-dm-events-2.

agrover commented:
Do you want to open a separate PR so I can review your latest changes?

agrover commented:
rebased.
